### ISSUE TITLE
Annotations are ignored when comparing overridden method with [NonNullTypes(false)]

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#28684]
No warnings are reported for `B.F` and `B.G` even though the annotations do not match the overridden methods in `A`.

See `Overriding_23`:

```c#
using System.Runtime.CompilerServices;
namespace System.Runtime.CompilerServices
{
    public sealed class NonNullTypesAttribute : Attribute
    {
        public NonNullTypesAttribute(bool flag) { }
    }
}
[NonNullTypes(true)]
abstract class A
{
    internal abstract string F();
    internal abstract object?[] G();
}
[NonNullTypes(false)]
abstract class B : A
{
    internal override abstract string? F(); // does not match string
    internal override abstract object[] G(); // does not match object?[]
}
```

**Comment by cston** [comment]
cc @jcouv

**Comment by AlekseyTs** [comment]
Here is another scenario:
```

        [Fact]
        public void ImplementingNonNullWithNullable()
        {
            var source = @"
interface IA
{
    string[] M1();
    T[] M2<T>() where T : class;
}

class B : IA
{
    [System.Runtime.CompilerServices.NonNullTypes(false)]
    string?[] IA.M1()
    {
        return new string?[] {};
    }

    [System.Runtime.CompilerServices.NonNullTypes(false)]
    S?[] IA.M2<S>()
    {
        return new S?[] {};
    }
}
";
            // PROTOTYPE(NullableReferenceTypes): missing a warning on IA.M2
            var compilation = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
            compilation.VerifyDiagnostics(
                // (11,18): warning CS8616: Nullability of reference types in return type doesn't match implemented member 'string[] IA.M1()'.
                //     string?[] IA.M1()
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInReturnTypeOnExplicitImplementation, "M1").WithArguments("string[] IA.M1()").WithLocation(11, 18),
                // (17,6): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //     S?[] IA.M2<S>()
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(17, 6),
                // (11,11): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //     string?[] IA.M1()
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(11, 11),
                // (13,26): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //         return new string?[] {};
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(13, 26),
                // (19,21): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //         return new S?[] {};
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(19, 21)
                );
        }
```

And one more:
```
        [Fact]
        public void Overriding_23()
        {
            var source = @"
class C
{
    public static void Main()
    { 
    }
}

abstract class A
{
    public abstract string[] M1(); 
    public abstract T[] M2<T>() where T : class; 
}

class B : A
{
    [System.Runtime.CompilerServices.NonNullTypes(false)]
    public override string?[] M1()
    {
        return new string?[] {};
    } 

    [System.Runtime.CompilerServices.NonNullTypes(false)]
    public override S?[] M2<S>()
    {
        return new S?[] {};
    } 
}
";
            var compilation = CreateCompilation(new[] { source, NonNullTypesTrue, NonNullTypesAttributesDefinition });
            // https://github.com/dotnet/roslyn/issues/28684: Should report return type mismatch
            // for M2 as well.
            compilation.VerifyDiagnostics(
                // (18,31): warning CS8609: Nullability of reference types in return type doesn't match overridden member.
                //     public override string?[] M1()
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInReturnTypeOnOverride, "M1").WithLocation(18, 31),
                // (24,22): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //     public override S?[] M2<S>()
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(24, 22),
                // (18,27): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //     public override string?[] M1()
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(18, 27),
                // (20,26): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //         return new string?[] {};
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(20, 26),
                // (26,21): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //         return new S?[] {};
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(26, 21)
                );
        }
```

**Comment by AlekseyTs** [comment]
One more:
```
        [Fact]
        public void Overriding_24()
        {
            var source = @"
[module: System.Runtime.CompilerServices.NonNullTypes(true)]

abstract class A
{
    public abstract void M1(string[] x); 
    public abstract void M2<T>(T[] x) where T : class; 
}

class B : A
{
    [System.Runtime.CompilerServices.NonNullTypes(false)]
    public override void M1(string?[] x)
    {
    } 

    [System.Runtime.CompilerServices.NonNullTypes(false)]
    public override void M2<T>(T?[] x)
    {
    } 
}
";
            var compilation = CreateCompilation(new[] { source, NonNullTypesAttributesDefinition });

            compilation.VerifyDiagnostics(
                // (13,26): warning CS8610: Nullability of reference types in type of parameter 'x' doesn't match overridden member.
                //     public override void M1(string?[] x)
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInParameterTypeOnOverride, "M1").WithArguments("x").WithLocation(13, 26),
                // (18,33): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //     public override void M2<T>(T?[] x)
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(18, 33),
                // (13,35): warning CS8632: The annotation for nullable reference types should only be used in code within a '[NonNullTypes(true)]' context.
                //     public override void M1(string?[] x)
                Diagnostic(ErrorCode.WRN_MissingNonNullTypesContextForAnnotation, "?").WithLocation(13, 35)
                );
            // PROTOTYPE(NullableReferenceTypes): should warn WRN_NullabilityMismatchInParameterTypeOnOverride on B.M2 too
        }
```
